src/analysis/validation.jl

```diff
@@ -76,88 +67,27 @@ function get_tank_dataframe(wm_data,wm_solution,wntr_data,wntr_simulation, tank_
-        link_type = "pipe"
-    elseif link_id in keys(wm_data["nw"]["1"]["short_pipe"])
-        link_type = "short_pipe"
-    elseif link_id in keys(wm_data["nw"]["1"]["valve"])
-        link_type = "valve"
-    elseif link_id in keys(wm_data["nw"]["1"]["regulator"])
-        link_type = "regulator"
-    else
-        link_type = "pump"
-    end
-
-    num_time_step = length(wm_solution["solution"]["nw"])            # number of time steps
-    time_step = wm_data["time_step"]/3600                       # length per time step (hour)
-    link_name = string(wm_data["nw"]["1"][link_type][link_id]["source_id"][2])
-    flow_wntr = Array{Float64,1}(undef,num_time_step)
-    flow_watermodels = Array{Float64,1}(undef,num_time_step)
-    head_loss_wntr = Array{Float64,1}(undef,num_time_step)
-    head_loss_watermodels = Array{Float64,1}(undef,num_time_step)
-    start_node_id = wm_data["nw"]["1"][link_type][link_id]["node_fr"]
-    end_node_id = wm_data["nw"]["1"][link_type][link_id]["node_to"]
-
-    # if this is a shutoff valve next to a tank, link_name should include prefix
-    for (key,tank) in wm_data["nw"]["1"]["tank"]
-        if end_node_id == wm_data["nw"]["1"]["tank"][key]["node"]   # if a link's end node is a tank, this link MUST be an added shutoff valve
-            link_name = "atl" * link_name
-            break
-        end
-    end
-    # similarly, if this is a zero-length check valve (artificial), link_name should include prefix
-    if link_type == "valve"
-        if wm_data["nw"]["1"][link_type][link_id]["source_id"][1] == "pipe"
-            link_name = "avl" * link_name
-        end
-    end
-
-    start_node_name = wntr_data.links._data[link_name].start_node_name
-    end_node_name = wntr_data.links._data[link_name].end_node_name
-
-    # Obtain relevant WNTR solution data.
-    flow_wntr = _get_wntr_link_attribute(wntr_simulation, link_name, "flowrate")
-    head_start_wntr = _get_wntr_node_attribute(wntr_simulation, start_node_name, "head")
-    head_end_wntr = _get_wntr_node_attribute(wntr_simulation, end_node_name, "head")
-    head_loss_wntr = head_start_wntr .- head_end_wntr
-
-    for t in 1:num_time_step
-        # Calculate WaterModels solution data.
-        flow_watermodels[t] = wm_solution["solution"]["nw"][string(t)][link_type][link_id]["q"]
-        head_loss_watermodels[t] = wm_solution["solution"]["nw"][string(t)]["node"][string(start_node_id)]["h"]-wm_solution["solution"]["nw"][string(t)]["node"][string(end_node_id)]["h"]
-    end
-
-    link_df = DataFrames.DataFrame(time = 1:time_step:time_step*num_time_step, flow_wntr = flow_wntr, flow_watermodels = flow_watermodels, 
-        head_loss_wntr = head_loss_wntr, head_loss_watermodels = head_loss_watermodels)
-
-    return link_df
-end
 
 # compare pipes
 function get_pipe_dataframe(wm_data,wm_solution,wntr_data,wntr_simulation,pipe_id)
     num_time_step = length(wm_solution["solution"]["nw"])            # number of time steps
```

Comment:
There is a lot of duplicated code blocks for each of these functions `get_pipe_dataframe`, `get_short_pipe_dataframe`, and `get_valve_dataframe`. You might consider making a function (or two) that could execute these blocks with a keyword to designate the link type. Just a suggestion.